etc/make_small_lib.pl

```diff
@@ -0,0 +1,335 @@
+  }
+  $tcpath = $path . $sep . 'tommath_class.h';
+  write_file($tcpath, $content);
+  warn "File $tcpath written\n";
+  $tcpath = $path . $sep . 'tommath_superclass.h';
+  write_file($tcpath, "");
+  warn "File $tcpath emptied\n";
+}
+
+sub gather_functions
+{
+  my $path = shift;
+  my %depmap;
+
+  if(-d $path) {
+    $path = $path . $sep . '*.c';
+  }
+
+  foreach my $filename (glob $path) {
+print "filename: " . $filename . " path: ". $path . "\n";
+    open(my $src, '<', $filename) or die "Can't open source file!\n";
+    read $src, my $content, -s $src;
+    close $src;
+
+    $filename = basename($filename);
+    $filename =~ s/^bn_|\.c$//g;
+
+    $content =~ s{/\*.*?\*/}{}gs;
+    my $list = "";
+    foreach my $line (split /\n/, $content) {
+      # TODO: change to read public functions only when scanning user source.
+      while ($line =~ /(fast_)?(s_)?mp\_[a-z_0-9]*(?=\()|(?<=\()mp\_[a-z_0-9]*(?=,)/g) {
+        my $a = $&;
+        next if $a eq "mp_err";
+        if($list eq "") {
+          $list = $a;
+        }
+        else {
+          $list = $list . "," . $a;
+        }
+      }
+    }
+    $depmap{$filename} = $list;
+  }
+  return %depmap;
+}
+
+sub gather_dependencies
+{
+   my ($deplist, $depmap, $funcslist) = @_;
+   my @funcs = split ',', $funcslist;
+   if ($deplist =~ /$funcs[0]/) {
+      return $deplist;
+   } else {
+      $deplist = $deplist . $funcs[0];
+   }
+   push @dependency_list, $funcs[0];
+   shift @funcs;
+   my $olddeplist = $deplist;
+   foreach my $i (@funcs) {
+      $deplist = gather_dependencies($deplist, $depmap, ${$depmap}{$i}) if exists ${$depmap}{$i};
+   }
+   return $olddeplist;
+}
+
+sub start
+{
+  my ($sd, $td, $no, $cm) = @_;
+
+  my %depmap;
+  my %user_functions;
+  my %ff_hash;
+  my @functions;
+  my @tmp;
+
+  # TODO: checks&balances
+   -e $td.$sep."tommath.h" or die "$td.$sep.tommath.h not found, please check path to LibTomMath sources\n";
+
+  %depmap = gather_functions($td);
+  %user_functions = gather_functions($sd);
+
+  foreach (sort keys %user_functions) {
+    push @functions, split /,/ , $user_functions{$_};
+  }
+  @functions = uniq(sort @functions);
+
+  foreach (@functions) {
+    exists $depmap{$_} or die "Function \"$_\" does not exist in LibTomMath.\n";
+  }
+
+  # gather dependencies
+  foreach (sort keys %user_functions) {
+    gather_dependencies("", \%depmap, $user_functions{$_});
+  }
+  @dependency_list = uniq(sort @dependency_list);
+
+  # make an even smaller lib by removing non-essential functions (e.g. Karatsuba for multiplication)
+  if ($no == 1) {
+    foreach (@dependency_list) {
+      next if exists $fast_functions{$_};
+      push @tmp, $_;
+    }
+    @dependency_list = @tmp;
+    @tmp = ();
+  }
+
+  if ($cm == 1) {
+    foreach my $entry (@dependency_list) {
+      $entry = 'bn_' . $entry . '.o';
+      push @tmp, $entry;
+    }
+    process_makefiles($td, @tmp);
+  }
+  else {
+    write_header($td, @dependency_list);
+  }
```

Comment:
I would prefer if there is a single --update commandline, and then both makefiles and tommath_config.h header are rewritten.

Furthermore I would love to have an amalgam being generated too, but I could also contribute this at some later point :grinning: It would replace gen.pl

Reply:
> I would prefer if there is a single --update commandline, and then both makefiles and tommath_config.h header are rewritten.

You need only one to make it work, but is is of course no problem to remove this condition and allow both options simultaneously.

> Furthermore I would love to have an amalgam being generated too

Give them an inch and they'll take an ell. ;-)

I'd rather change `gen.pl` to read from `tommath_class.h` which has either everything listed between `LTM_ALL` and `endif` or contains the reduced set generated by this script. It would also be more "Unix-y".
That would ignore any manually made changes in `tommath_superclass.h`, of course, but the current `gen.pl` ignores them, too.

> Since you still have the depmap, maybe you could also import the callgraph logic from helper.pl?

In one comment you ask: "Could we avoid the over configuration?" and here you want something extra? ;-)

The callgraph logic from `helper.pl`s is already used in this script, didn't want to reinvent the wheel.